File: 04/solution.py
#! /usr/bin/env python3

with open('input.txt', 'r') as file:
    input = file.read().rstrip()

# input = '''
# .M.S......
# ..A..MSMS.
# .M.S.MAA..
# ..A.ASMSM.
# .M.S.M....
# ..........
# S.S.S.S.S.
# .A.A.A.A..
# M.M.M.M.M.
# ..........
# '''

grid = [list(line) for line in input.splitlines() if line]
WIDTH, HEIGHT = len(grid[0]), len(grid)

# part 1

# find all 'X' positions to start
x_pos = []
for x in range(WIDTH):
    for y in range(HEIGHT):
        if grid[y][x] == 'X':
            x_pos.append((x, y, 'X', None))

NEXT_CHAR = {'X': 'M', 'M': 'A', 'A': 'S'}
ADJ = [
    (-1, -1), (0, -1), (1, -1),
    (-1,  0),          (1,  0),
    (-1,  1), (0,  1), (1,  1),
]

def get_valid_neighbours(pos):
    x, y, char, dir = pos

    neighbours = []

    # Keep current direction
    if dir is None:
        adj = ADJ
    else:
        adj = [dir]

    for (dx, dy) in adj:
        nx, ny = x + dx, y + dy
        if nx < 0 or nx >= WIDTH or ny < 0 or ny >= HEIGHT:
            continue

        new_char = grid[ny][nx]
        if new_char == NEXT_CHAR[char]:
            neighbours.append((nx, ny, new_char, (dx, dy)))

    return neighbours

total_xmas = 0

seen = set()
q = [pos for pos in x_pos]
# q = [pos for pos in x_pos[0:1]]
while len(q):
    pos = q.pop(0)
    neighbours = get_valid_neighbours(pos)
    for n in neighbours:
        _, _, char, _ = n
        if char == 'S':
            total_xmas += 1
        else:
            q.append(n)


print(total_xmas) # 2358

# part 2

# X-MAS

# find all 'A' positions to start
a_pos = []
for x in range(WIDTH):
    for y in range(HEIGHT):
        if grid[y][x] == 'A':
            a_pos.append((x, y, 'A'))

# get 'X' shape around
ADJ = [
    (-1, -1), (1, -1),
    (-1,  1), (1,  1),
]

x_mas_count = 0

valid_strs = set([
    'SSMM',
    'MSMS',
    'SMSM',
    'MMSS',
])

for a in a_pos:
    x, y, char = a

    invalid = False
    ns = []
    for (dx, dy) in ADJ:
        nx, ny = x + dx, y + dy
        if nx < 0 or nx >= WIDTH or ny < 0 or ny >= HEIGHT:
            invalid = True
            break
        new_char = grid[ny][nx]
        if new_char in ('M', 'S'):
            ns.append((nx, ny, new_char))
    if invalid:
        continue

    # two mas in shape of an x

    c_s, c_m = 0, 0
    str = ''
    for n in ns:
        _, _, char = n
        if char == 'S':
            c_s += 1
        elif char == 'M':
            c_m += 1

        str += (char)

    if str in valid_strs:
        x_mas_count += 1

print(x_mas_count) # 1737

File: 04/solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

var (
	grid          [][]rune
	WIDTH, HEIGHT int
	NEXT_CHAR     = map[rune]rune{'X': 'M', 'M': 'A', 'A': 'S'}
	// Neighbours (Square)
	ADJ = [][2]int{
		{-1, -1}, {0, -1}, {1, -1},
		{-1, 0}, {1, 0},
		{-1, 1}, {0, 1}, {1, 1},
	}
	// Neighbours (X-shape)
	ADJ_X = [][2]int{
		{-1, -1}, {1, -1},
		{-1, 1}, {1, 1},
	}
	VALID_NEIGHBOUR_STRINGS = map[string]bool{
		"SSMM": true,
		"MSMS": true,
		"SMSM": true,
		"MMSS": true,
	}
)

func main() {
	file, err := os.Open("input.txt")
	if err != nil {
		panic(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		if len(line) > 0 {
			grid = append(grid, []rune(line))
		}
	}

	HEIGHT, WIDTH = len(grid), len(grid[0])

	// Part 1
	totalXmas := part1()
	fmt.Println(totalXmas) // 2358

	// Part 2
	xMasCount := part2()
	fmt.Println(xMasCount) // 1737
}

func part1() int {
	xPositions := findPositions('X')
	queue := make([][4]interface{}, len(xPositions))
	copy(queue, xPositions)
	totalXmas := 0

	for len(queue) > 0 {
		pos := queue[0]
		queue = queue[1:]
		neighbours := getValidNeighbours(pos)

		for _, n := range neighbours {
			if n[2].(rune) == 'S' {
				totalXmas++
			} else {
				queue = append(queue, n)
			}
		}
	}

	return totalXmas
}

func part2() int {
	aPositions := findPositions('A')
	xMasCount := 0

	for _, pos := range aPositions {
		x, y := pos[0].(int), pos[1].(int)
		neighbourStr := ""

		for _, adj := range ADJ_X {
			nx, ny := x+adj[0], y+adj[1]
			if nx >= 0 && nx < WIDTH && ny >= 0 && ny < HEIGHT {
				char := grid[ny][nx]
				if char == 'M' || char == 'S' {
					neighbourStr += string(char)
				}
			}
		}

		if len(neighbourStr) == 4 && VALID_NEIGHBOUR_STRINGS[neighbourStr] {
			xMasCount++
		}
	}

	return xMasCount
}

func findPositions(target rune) [][4]interface{} {
	positions := [][4]interface{}{}
	for y := 0; y < HEIGHT; y++ {
		for x := 0; x < WIDTH; x++ {
			if grid[y][x] == target {
				positions = append(positions, [4]interface{}{x, y, target, nil})
			}
		}
	}
	return positions
}

func getValidNeighbours(pos [4]interface{}) [][4]interface{} {
	x, y := pos[0].(int), pos[1].(int)
	char := pos[2].(rune)
	dir := pos[3]

	neighbours := [][4]interface{}{}
	adj := ADJ
	if dir != nil {
		d := dir.([2]int)
		adj = [][2]int{d}
	}

	for _, d := range adj {
		nx, ny := x+d[0], y+d[1]
		if nx >= 0 && nx < WIDTH && ny >= 0 && ny < HEIGHT {
			newChar := grid[ny][nx]
			if newChar == NEXT_CHAR[char] {
				neighbours = append(neighbours, [4]interface{}{nx, ny, newChar, d})
			}
		}
	}

	return neighbours
}
